### PR TITLE
Колоколова Дарья. Задача 3. Вариант 19. Поразрядная сортировка для целых чисел с четно-нечетным слиянием Бэтчера.

### PR DESCRIPTION
**Pull request by dasha1112**
Описание последовательной задачи
В последовательной версии реализована поразрядная сортировка для целых чисел с помощью функции radix_sort. Сначала она находит максимальное и минимальное значения в векторе, чтобы определить диапазон разрядов для сортировки. Затем, с помощью цикла, она вызывает функцию counting_sort_radix для сортировки чисел по разрядам, начиная с наименьшего. После обработки всех разрядов массив делится на два вектора: один для отрицательных чисел и другой для положительных. Отрицательные числа сортируются по возрастанию, а положительные остаются в отсортированном порядке. В конце оба вектора объединяются, и возвращается итоговый отсортированный массив.

Описание MPI задачи
В MPI версии реализован алгоритм поразрядной сортировки с четно-нечетным слиянием Бэтчера согласно его логике работы. Сначала на вход подается вектор целых чисел, после чего он делится между процессами. Далее в цикле, который повторяется ровно столько раз, сколько всего было запущено процессов, каждый процесс обменивается данными с соседним процессом, согласно логике алгоритма сортировки четно-нечетным слиянием Бэтчера. То есть, если итерация кратна двум, то процессы с четными рангами отправляют соседнему процессу свою часть вектора, а те в свою очередь при помощи функции radix_sort сортируют вектор, полученный при помощи слияния этих двух процессов. Далее процессу с меньшим рангом отправляется соответственно часть вектора с минимальными значениями, а процессу с большим рангом достается часть вектора с максимальными значениями. Аналогичная ситуация происходит на итерации не кратной, но уже с процессами с нечетными рангами. После цикла со всех процессов собираются результаты с помощью gather.

### CODE REVIEW

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  if (world.rank() == 0) {
+    taskDataPar->inputs.emplace_back(reinterpret_cast<uint8_t *>(unsorted_vector.data()));
+    taskDataPar->inputs_count.emplace_back(unsorted_vector.size());
+    taskDataPar->outputs.emplace_back(reinterpret_cast<uint8_t *>(sorted_vector.data()));
+    taskDataPar->outputs_count.emplace_back(sorted_vector.size());
+  }
+
+  if (world.rank() == 0) {
+    kolokolova_d_radix_integer_merge_sort_mpi::TestMPITaskParallel testMpiTaskParallel(taskDataPar);
+    ASSERT_EQ(testMpiTaskParallel.validation(), false);
```
Comment (Kitos87) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:188]: Please, use ASSERT_FALSE

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    kolokolova_d_radix_integer_merge_sort_mpi::TestMPITaskSequential testTaskSequential(taskDataSeq);
+
+    ASSERT_EQ(testTaskSequential.validation(), true);
+    testTaskSequential.pre_processing();
+    testTaskSequential.run();
+    testTaskSequential.post_processing();
+
+    for (int i = 0; i < size_vector; i++) {
+      ASSERT_EQ(sorted_vector[i], reference_res[i]);
```
Comment (Kitos87) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:164]: Please, use EXPECT_EQ

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    // Create TaskData
+    std::shared_ptr<ppc::core::TaskData> taskDataSeq = std::make_shared<ppc::core::TaskData>();
+    taskDataSeq->inputs.emplace_back(reinterpret_cast<uint8_t *>(unsorted_vector.data()));
+    taskDataSeq->inputs_count.emplace_back(unsorted_vector.size());
+    taskDataSeq->outputs.emplace_back(reinterpret_cast<uint8_t *>(reference_res.data()));
+    taskDataSeq->outputs_count.emplace_back(reference_res.size());
+
+    kolokolova_d_radix_integer_merge_sort_mpi::TestMPITaskSequential testTaskSequential(taskDataSeq);
+
+    ASSERT_EQ(testTaskSequential.validation(), true);
```
Comment (Kitos87) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:158]: Please, use ASSERT_TRUE

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  if (world.rank() == 0) {
+    unsorted_vector = kolokolova_d_radix_integer_merge_sort_mpi::getRandomVector(size_vector);
+    taskDataPar->inputs.emplace_back(reinterpret_cast<uint8_t *>(unsorted_vector.data()));
+    taskDataPar->inputs_count.emplace_back(unsorted_vector.size());
+    taskDataPar->outputs.emplace_back(reinterpret_cast<uint8_t *>(sorted_vector.data()));
+    taskDataPar->outputs_count.emplace_back(sorted_vector.size());
+  }
+
+  kolokolova_d_radix_integer_merge_sort_mpi::TestMPITaskParallel testMpiTaskParallel(taskDataPar);
+  ASSERT_EQ(testMpiTaskParallel.validation(), true);
```
Comment (Kitos87) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:140]: Please, use ASSERT_TRUE

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+    kolokolova_d_radix_integer_merge_sort_mpi::TestMPITaskSequential testTaskSequential(taskDataSeq);
+
+    ASSERT_EQ(testTaskSequential.validation(), true);
+    testTaskSequential.pre_processing();
+    testTaskSequential.run();
+    testTaskSequential.post_processing();
+
+    for (int i = 0; i < size_vector; i++) {
+      ASSERT_EQ(sorted_vector[i], reference_res[i]);
```
Comment (Kitos87) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:116]: Please, use EXPECT_EQ

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    // Create TaskData
+    std::shared_ptr<ppc::core::TaskData> taskDataSeq = std::make_shared<ppc::core::TaskData>();
+    taskDataSeq->inputs.emplace_back(reinterpret_cast<uint8_t *>(unsorted_vector.data()));
+    taskDataSeq->inputs_count.emplace_back(unsorted_vector.size());
+    taskDataSeq->outputs.emplace_back(reinterpret_cast<uint8_t *>(reference_res.data()));
+    taskDataSeq->outputs_count.emplace_back(reference_res.size());
+
+    kolokolova_d_radix_integer_merge_sort_mpi::TestMPITaskSequential testTaskSequential(taskDataSeq);
+
+    ASSERT_EQ(testTaskSequential.validation(), true);
```
Comment (Kitos87) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:110]: Please, use ASSERT_TRUE

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  if (world.rank() == 0) {
+    unsorted_vector = kolokolova_d_radix_integer_merge_sort_mpi::getRandomVector(size_vector);
+    taskDataPar->inputs.emplace_back(reinterpret_cast<uint8_t *>(unsorted_vector.data()));
+    taskDataPar->inputs_count.emplace_back(unsorted_vector.size());
+    taskDataPar->outputs.emplace_back(reinterpret_cast<uint8_t *>(sorted_vector.data()));
+    taskDataPar->outputs_count.emplace_back(sorted_vector.size());
+  }
+
+  kolokolova_d_radix_integer_merge_sort_mpi::TestMPITaskParallel testMpiTaskParallel(taskDataPar);
+  ASSERT_EQ(testMpiTaskParallel.validation(), true);
```
Comment (Kitos87) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:92]: Please, use ASSERT_TRUE

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
```
Comment (Kitos87) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:44]: Please, use ASSERT_TRUE

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    // Create TaskData
+    std::shared_ptr<ppc::core::TaskData> taskDataSeq = std::make_shared<ppc::core::TaskData>();
+    taskDataSeq->inputs.emplace_back(reinterpret_cast<uint8_t *>(unsorted_vector.data()));
+    taskDataSeq->inputs_count.emplace_back(unsorted_vector.size());
+    taskDataSeq->outputs.emplace_back(reinterpret_cast<uint8_t *>(reference_res.data()));
+    taskDataSeq->outputs_count.emplace_back(reference_res.size());
+
+    kolokolova_d_radix_integer_merge_sort_mpi::TestMPITaskSequential testTaskSequential(taskDataSeq);
+
+    ASSERT_EQ(testTaskSequential.validation(), true);
```
Comment (Kitos87) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:62]: Same

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    testTaskSequential.pre_processing();
+    testTaskSequential.run();
+    testTaskSequential.post_processing();
+
+    for (int i = 0; i < size_vector; i++) {
+      ASSERT_EQ(sorted_vector[i], reference_res[i]);
```
Comment (Kitos87) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:68]: Please, use EXPECT_EQ

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp (hunk 1 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp]: Fixed

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp (hunk 2 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp]: Fixed

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp (hunk 4 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp]: Fixed

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp (hunk 5 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp]: Fixed

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp (hunk 6 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp]: Fixed

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp (hunk 7 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp]: Fixed

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp (hunk 8 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp]: Fixed

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp (hunk 9 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp]: Fixed

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp (hunk 10 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp]: Fixed

**Review by Kitos87 — APPROVED** [review]

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    taskDataPar->inputs_count.emplace_back(unsorted_vector.size());
+    taskDataPar->outputs.emplace_back(reinterpret_cast<uint8_t *>(sorted_vector.data()));
+    taskDataPar->outputs_count.emplace_back(sorted_vector.size());
+  }
+
+  if (world.rank() == 0) {
+    kolokolova_d_radix_integer_merge_sort_mpi::TestMPITaskParallel testMpiTaskParallel(taskDataPar);
+    ASSERT_FALSE(testMpiTaskParallel.validation());
+  }
+}
```
Comment (antihype320) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:190]: Please, add test when output is empty to check validation

Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:213]: Fixing

**Review by antihype320 — APPROVED** [review]

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/include/ops_mpi.hpp [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+#include <numeric>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+
+namespace kolokolova_d_radix_integer_merge_sort_mpi {
+
+std::vector<int> getRandomVector(int sz);
```
Comment (aobolensk) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/include/ops_mpi.hpp:17]: Please, move test function to the test module

File: tasks/seq/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 13]
```diff
@@ -0,0 +1,90 @@
+// Copyright 2023 Nesterov Alexander
+#include <gtest/gtest.h>
+
+#include <vector>
+
+#include "seq/kolokolova_d_radix_integer_merge_sort/include/ops_seq.hpp"
+
```
Comment (aobolensk) [inline tasks/seq/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:7]: Please, add random tests

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/perf_tests/main.cpp [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+using namespace kolokolova_d_radix_integer_merge_sort_mpi;
+
+std::vector<int> kolokolova_d_radix_integer_merge_sort_mpi::getRandomVector(int sz) {
+  std::random_device dev;
+  std::mt19937 gen(dev());
+  std::vector<int> vec(sz);
+  std::uniform_int_distribution<int> dist(-100, 100);
+  for (int i = 0; i < sz; i++) {
+    vec[i] = gen() % 100;
```
Comment (aobolensk) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/perf_tests/main.cpp:21]: Please, use negative numbers

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+
+#include "mpi/kolokolova_d_radix_integer_merge_sort/include/ops_mpi.hpp"
+
+using namespace kolokolova_d_radix_integer_merge_sort_mpi;
+
+std::vector<int> kolokolova_d_radix_integer_merge_sort_mpi::getRandomVector(int sz) {
+  std::random_device dev;
+  std::mt19937 gen(dev());
+  std::vector<int> vec(sz);
+  std::uniform_int_distribution<int> dist(-10000, 10000);
```
Comment (allnes) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:18]: please parametrize this constants in tests

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+  for (int i = 0; i < sz; i++) {
+    vec[i] = dist(gen);
+  }
+  return vec;
+}
+
+TEST(kolokolova_d_radix_integer_merge_sort_mpi, Test_Parallel_Sort1) {
+  boost::mpi::communicator world;
+  int size_vector = world.size() * 4;
```
Comment (allnes) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:27]: why do you do multiplication?

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/src/ops_mpi.cpp [hunk 17]
```diff
@@ -0,0 +1,217 @@
+#include "mpi/kolokolova_d_radix_integer_merge_sort/include/ops_mpi.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <random>
+#include <string>
+#include <thread>
+#include <vector>
+
+using namespace std::chrono_literals;
```
Comment (allnes) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/src/ops_mpi.cpp:10]: please remove it

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/src/ops_mpi.cpp [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+
+void kolokolova_d_radix_integer_merge_sort_mpi::counting_sort_radix(std::vector<int>& array, int exp) {
+  int size_vector = int(array.size());
+  std::vector<int> func_res(size_vector);
+  std::vector<int> nums_of_digits(20, 0);
+
+  for (int i = 0; i < size_vector; i++) {
+    int index = (array[i] / exp) % 10;
```
Comment (allnes) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/src/ops_mpi.cpp:18]: exp is reservated name

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/src/ops_mpi.cpp [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+  for (int i = 0; i < int(sort_vector.size()); i++) {
+  }
+  return sort_vector;
+}
+
+std::vector<int> kolokolova_d_radix_integer_merge_sort_mpi::radix_sort(std::vector<int>& array) {
+  int max_num = *max_element(array.begin(), array.end());
+  int min_num = *min_element(array.begin(), array.end());
+
+  for (int exp = 1; max_num / exp > 0 || min_num / exp < 0; exp *= 10) {
```
Comment (allnes) [inline tasks/mpi/kolokolova_d_radix_integer_merge_sort/src/ops_mpi.cpp:57]: please describe this loop

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/include/ops_mpi.hpp (hunk 12 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/include/ops_mpi.hpp]: Fixed

File: tasks/seq/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp (hunk 13 above)
Comment (dasha1112) [reply tasks/seq/kolokolova_d_radix_integer_merge_sort/func_tests/main.cpp:10]: Fixed

File: tasks/mpi/kolokolova_d_radix_integer_merge_sort/perf_tests/main.cpp (hunk 14 above)
Comment (dasha1112) [reply tasks/mpi/kolokolova_d_radix_integer_merge_sort/perf_tests/main.cpp]: Fixed

**Review by aobolensk — APPROVED** [review]
